src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,749 @@
+package is.hail.expr.ir
+
+import is.hail.HailContext
+import is.hail.annotations.BroadcastRow
+import is.hail.expr._
+import is.hail.expr.types._
+import is.hail.utils._
+import org.apache.spark.sql.Row
+
+object PruneDeadFields {
+  def isSupertype(superType: BaseType, subType: BaseType): Boolean = {
+    try {
+      (superType, subType) match {
+        case (tt1: TableType, tt2: TableType) =>
+          isSupertype(tt1.globalType, tt2.globalType) &&
+            isSupertype(tt1.rowType, tt2.rowType)
+        case (mt1: MatrixType, mt2: MatrixType) =>
+          isSupertype(mt1.globalType, mt2.globalType) &&
+            isSupertype(mt1.rowType, mt2.rowType) &&
+            isSupertype(mt1.colType, mt2.colType) &&
+            isSupertype(mt1.entryType, mt2.entryType)
+        case (TArray(et1, r1), TArray(et2, r2)) => r1 == r2 && isSupertype(et1, et2)
+        case (TSet(et1, r1), TSet(et2, r2)) => r1 == r2 && isSupertype(et1, et2)
+        case (TDict(kt1, vt1, r1), TDict(kt2, vt2, r2)) => r1 == r2 && isSupertype(kt1, kt2) && isSupertype(vt1, vt2)
+        case (s1: TStruct, s2: TStruct) =>
+          var idx = -1
+          s1.required == s2.required && s1.fields.forall { f =>
+            val s2field = s2.field(f.name)
+            if (s2field.index > idx) {
+              idx = s2field.index
+              isSupertype(f.typ, s2field.typ)
+            } else
+              false
+          }
+        case (agg1: TAggregable, agg2: TAggregable) => isSupertype(agg1.elementType, agg2.elementType)
+        case (t1: Type, t2: Type) => t1 == t2
+        case _ => fatal(s"invalid comparison: $superType / $subType")
+      }
+    } catch {
+      case e: Throwable =>
+        fatal(s"error while checking subtype:\n  super: $superType\n  sub:   $subType", e)
+    }
+  }
+
+  def apply(ir: BaseIR): BaseIR = {
+
+    try {
+      ir.deepCopy() match {
+        case x: MatrixIR => rebuild(x, x.typ)
+        case x: TableIR => rebuild(x, x.typ)
+        case x: IR => rebuild(x)
+      }
+    } catch {
+      case e: Throwable => fatal(s"error trying to rebuild IR:\n${ Pretty(ir) }", e)
+    }
+  }
+
+
+  def minimal(tt: TableType): TableType = {
```

Comment:
I think I'd write this as `def minimalBaseType(t: BaseType): BaseType` and then `def minimal[T <: BaseType](t: T): T = minimalBaseType(t).asInstanceOf[T]`.